### PR TITLE
libmisc, man: Drop old check and advice for complex character sets in passwords

### PR DESCRIPTION
**Pull request by alejandro-colomar**
I kept the minimum length as 8, since it was what we had for the simplest passwords already.  8 is probably not what you really want from a password, but I believe it should be up to users to chose what they want.  If we raise this too much, we may upset some people, so let's be cautious.

Add the relevant XKCD to the passwd(1) manual page.  It already explains most of the rationale behind this patch.

The strength of a password is an exponential formula, where the base is the size of the character set, and the exponent is the length of the password.  That already shows why long passwords of just lowercase letters are better than short Pa$sw0rdZ3.  But an even more important point is that humans, when forced to use symbols in a password, are more likely to do trivial substitutions on simple passwords, which doesn't increase strength, and can instead give a false sense of strength, which is dangerous.

Closes: <https://github.com/shadow-maint/shadow/issues/688>
Link: <https://xkcd.com/936/>

### CODE REVIEW

**Comment by alejandro-colomar** [comment]
Changes:

- wfix

**Comment by alejandro-colomar** [comment]
Add reference to makepasswd(1) in passwd(1)'s SEE ALSO.

File: man/passwd.1.xml [hunk 1]
```diff
@@ -488,5 +484,9 @@
 	<refentrytitle>usermod</refentrytitle><manvolnum>8</manvolnum>
       </citerefentry>.
     </para>
+
+    <para>
+	https://xkcd.com/936/
```
Comment (vapier) [inline man/passwd.1.xml:489]: this link lacks any sort of introduction or explanation

File: libmisc/obscure.c [hunk 2]
```diff
@@ -170,8 +119,8 @@ static /*@observer@*//*@null@*/const char *password_check (
 		msg = _("case changes only");
 	} else if (similar (oldmono, newmono)) {
 		msg = _("too similar");
-	} else if (simple (old, new)) {
-		msg = _("too simple");
+	} else if (strlen (new) < 8) {
```
Comment (vapier) [inline libmisc/obscure.c:122]: this is more restrictive than the existing code.  having it be dynamic based on the inputs was aligned with the intention of this code.  but you're making it static, and we already have `PASS_MIN_LEN` as a configurable value that is checked earlier.  so if you want to delete the dynamic strength checks, you should delete this length check entirely.

not that i have a problem with the dynamic strength check as-is.

File: man/passwd.1.xml [hunk 3]
```diff
@@ -94,23 +94,16 @@
       </para>
 
       <para>
-	Then, the password is tested for complexity. As a general guideline,
-	passwords should consist of 6 to 8 characters including one or more
-	characters from each of the following sets:
+	Then, the password is tested for length.  As a general guideline,
+	passwords should be long.  It's fine to use simple character sets,
+	such as passwords consisting only of lowercase letters.  For a
```
Comment (vapier) [inline man/passwd.1.xml:99]: this statement is much much too weak.  yes, it follows a "general guideline that passwords should be long", but this sentence feels disconnected from that and reads as independent guidance.

File: man/passwd.1.xml [hunk 4]
```diff
@@ -94,23 +94,16 @@
...
+	password consisting only of lowercase English letters randomly
+	chosen, and a length of 64, there are 27^64 (approximately 2^304)
```
Comment (vapier) [inline man/passwd.1.xml:101]: ok, but no one is generating & memorizing 64 randomly generated alphas.  it's one thing to use a password manager where long random generated values are never typed by users, but shadow is specifically geared towards users typing in their password interactively.

Comment (vapier) [inline man/passwd.1.xml:101]: you say "only lowercase English letters", but then say 27.  presumably you're including space in the count.

File: man/passwd.1.xml (hunk 1 above)
Comment (alejandro-colomar) [reply man/passwd.1.xml:489]: Hmm.  Usually references in SEE ALSO, such as references to other pages, are not introduced.  If a user goes to that site, they'll know why it was referenced.  But I'll think of adding some intro to it, especially for those who don't know what XKCD is, and foreseeing a possible future in which it doesn't exist anymore and people forgot what it was.

File: libmisc/obscure.c (hunk 2 above)
Comment (alejandro-colomar) [reply libmisc/obscure.c:122]: > this is more restrictive than the existing code. having it be dynamic based on the inputs was aligned with the intention of this code. but you're making it static, and we already have `PASS_MIN_LEN` as a configurable value that is checked earlier. so if you want to delete the dynamic strength checks, you should delete this length check entirely.

Yep, I didn't see the separate `PASS_MIN_LEN` test.  Then I'll just remove this test.

> 
> not that i have a problem with the dynamic strength check as-is.

It rewards passwords using wider character sets by allowing them to be shorter, which can reinforce the idea that it's a good thing to do in general.  Again, if one can remember long passwords using symbols and uppers, that's great, but enforcing that is problematic, and even this kind of reinforcement I believe it to be counterproductive in that regard.

File: man/passwd.1.xml (hunk 3 above)
Comment (alejandro-colomar) [reply man/passwd.1.xml:99]: True.  Maybe rephrased as: `It's fine to use simple character sets,
	such as passwords consisting only of lowercase letters` + ` if it helps memorize longer passwords`.

File: man/passwd.1.xml (hunk 4 above)
Comment (alejandro-colomar) [reply man/passwd.1.xml:101]: I do remember one, for my master PGP key.  Considering I only need a handful of passwords in my life (all the rest, are pass(1)'s task), I can memorize just a few very strong ones.  The trick I use to remember them is to break them into shorter chunks (say around 10 chars each), and treat them as words in a foreign language; I can certainly remember a few dozen "words" that are originally gibberish, with some repetition; then those words are connected to make sentences which are the actual password.

However, I admit my login, my SSH and my PGP signing passwords are half of that; not because I don't remember long ones, but because typing them dozens of times an hour would be cumbersome.  But 32 random `[:lower:][:digit:]` is definitely easy, and strong enough for almost anything.

Comment (alejandro-colomar) [reply man/passwd.1.xml:101]: Oops, miscount here :)  For some reason, `27 - 1 = 27` yesterday.

**Comment by alejandro-colomar** [comment]
Changes:

-  Completely drop test; don't replace it by a fixed length check.
-  Reword advice on passwd(1) to make the advice on simple char sets dependent on making longer passwords.
-  Introduce XKCD properly.

```
$ git range-diff master gh/strength strength 
1:  581b923a ! 1:  d101612b libmisc, man: Drop old check and advice for complex character sets in passwords
    @@ Metadata
      ## Commit message ##
         libmisc, man: Drop old check and advice for complex character sets in passwords
     
    -    I kept the minimum length as 8, since it was what we had for the
    -    simplest passwords already.  8 is probably not what you really want from
    -    a password, but I believe it should be up to users to chose what they
    -    want.  If we raise this too much, we may upset some people, so let's be
    -    cautious.
    -
         Add the relevant XKCD to the passwd(1) manual page.  It already explains
         most of the rationale behind this patch.
     
         Add also reference to makepasswd(1), which is a good way to generate
         strong passwords.  Personally, I commonly run `makepasswd --chars 64` to
    -    create my passwords.
    +    create my passwords, or 32 for passwords I need to type interactively
    +    often.
     
         The strength of a password is an exponential formula, where the base is
         the size of the character set, and the exponent is the length of the
    @@ libmisc/obscure.c: static /*@observer@*//*@null@*/const char *password_check (
                msg = _("too similar");
     -  } else if (simple (old, new)) {
     -          msg = _("too simple");
    -+  } else if (strlen (new) < 8) {
    -+          msg = _("too short");
        } else if (strstr (wrapped, newmono) != NULL) {
                msg = _("rotated");
        } else {
    @@ man/passwd.1.xml
     -  Then, the password is tested for complexity. As a general guideline,
     -  passwords should consist of 6 to 8 characters including one or more
     -  characters from each of the following sets:
    -+  Then, the password is tested for length.  As a general guideline,
    -+  passwords should be long.  It's fine to use simple character sets,
    -+  such as passwords consisting only of lowercase letters.  For a
    -+  password consisting only of lowercase English letters randomly
    -+  chosen, and a length of 64, there are 27^64 (approximately 2^304)
    -+  different possible combinations.  Being an exponential equation,
    -+  it's apparent that the exponent (the length) is more important than
    -+  the base (the size of the character set).
    -       </para>
    - 
    +-      </para>
    +-
     -      <itemizedlist mark='bullet'>
     -  <listitem>
     -    <para>lower case alphabetics</para>
    @@ man/passwd.1.xml
     -  </listitem>
     -      </itemizedlist>
     -
    +-      <para>
    +-  Care must be taken not to include the system default erase or kill
    +-  characters. <command>passwd</command> will reject any password which
    +-  is not suitably complex.
    ++  Then, the password is tested for complexity.
    ++  <command>passwd</command> will reject any password which is not
    ++  suitably complex.  Care must be taken not to include the system
    ++  default erase or kill characters.
    +       </para>
    + 
    +     </refsect2>
    +@@
    +   used as guesses to violate system security.
    +       </para>
    + 
    ++      <para>
    ++  As a general guideline, passwords should be long and random.  It's
    ++  fine to use simple character sets, such as passwords consisting
    ++  only of lowercase letters, if that helps memorizing longer
    ++  passwords.  For a password consisting only of lowercase English
    ++  letters randomly chosen, and a length of 32, there are 26^32
    ++  (approximately 2^150) different possible combinations.  Being an
    ++  exponential equation, it's apparent that the exponent (the length)
    ++  is more important than the base (the size of the character set).
    ++      </para>
    ++
            <para>
    -   Care must be taken not to include the system default erase or kill
    -   characters. <command>passwd</command> will reject any password which
    +   You can find advice on how to choose a strong password on
    +   http://en.wikipedia.org/wiki/Password_strength
     @@
            <citerefentry>
        <refentrytitle>chpasswd</refentrytitle><manvolnum>8</manvolnum>
    @@ man/passwd.1.xml
          </para>
     +
     +    <para>
    -+  https://xkcd.com/936/
    ++  The following web page comically (yet correctly) compares the
    ++  strength of two different methods for choosing a password:
    ++  "https://xkcd.com/936/"
     +    </para>
        </refsect1>
      </refentry>

```

**Comment by alejandro-colomar** [comment]
```
$ git range-diff master gh/strength strength 
1:  d101612b ! 1:  f9443f13 libmisc, man: Drop old check and advice for complex character sets in passwords
    @@ Commit message
     
         Closes: <https://github.com/shadow-maint/shadow/issues/688>
         Link: <https://xkcd.com/936/>
    +    Cc: Mike Frysinger <vapier@gentoo.org>
         Signed-off-by: Alejandro Colomar <alx@kernel.org>
     
      ## libmisc/obscure.c ##
```

**Comment by alejandro-colomar** [comment]
Changes:

- Rebased to master
- Sign commit with GPG

```
$ git range-diff f9443f13^..f9443f13 shadow/master..strength 
1:  f9443f13 = 1:  3fe545b2 libmisc, man: Drop old check and advice for complex character sets in passwords
```

**Review by hallyn — APPROVED** [review]

**Comment by hallyn** [comment]
@ikerexxe look good to you?

**Review by ikerexxe — APPROVED** [review]
Yes, it looks good to me. Merging!
